src/node/stage/mod.rs

```diff
@@ -11,337 +11,420 @@ mod bootstrapping;
+                }
+                Command::Transition(state) => {
+                    *self.state.lock().await = *state;
+                    Ok(vec![])
                 }
             }
         })
-        .await
+        .await;
+
+        if let Err(error) = &result {
+            error!("Error encountered when handling command: {}", error);
+        }
+
+        result
     }
 
-    pub async fn process_timeout(&mut self, token: u64) -> Result<()> {
-        log_ident::set(self.log_ident(), async {
-            match &mut self.state {
-                State::Bootstrapping(stage) => stage.process_timeout(token).await,
-                State::Joining(stage) => stage.process_timeout(token).await,
-                State::Approved(stage) => stage.process_timeout(token).await,
-            }
-        })
-        .await
-    }
-
-    // Checks whether the given location represents self.
-    async fn in_dst_location(&mut self, msg: &Message) -> Result<bool> {
-        let in_dst = match &mut self.state {
-            State::Bootstrapping(_) | State::Joining(_) => match msg.dst() {
-                DstLocation::Node(name) => name == &self.node_info().name(),
-                DstLocation::Section(_) => false,
-                DstLocation::Direct => true,
-            },
-            State::Approved(stage) => {
-                let is_dst_location = msg
-                    .dst()
-                    .contains(&stage.node_info.name(), stage.shared_state.our_prefix());
-
-                // Relay a message to the network if the message
-                // is not for us, or if it is for the section.
-                if !is_dst_location || msg.dst().is_section() {
-                    // Relay closer to the destination or
-                    // broadcast to the rest of our section.
-                    stage.relay_message(msg).await?;
-                }
+    // Note: this indirecton is needed. Trying to call `spawn(self.handle_commands(...))` directly
+    // inside `handle_commands` causes compile error about type check cycle.
+    fn spawn_handle_commands(self: Arc<Self>, command: Command) {
+        let _ = tokio::spawn(self.handle_commands(command));
+    }
 
-                is_dst_location
-            }
-        };
+    async fn handle_message(
+        &self,
+        sender: Option<SocketAddr>,
+        msg: Message,
+    ) -> Result<Vec<Command>> {
+        match &mut *self.state.lock().await {
+            State::Bootstrapping(state) => state.handle_message(sender, msg),
+            State::Joining(state) => state.handle_message(sender, msg),
+            State::Approved(state) => state.handle_message(sender, msg),
+        }
+    }
 
-        Ok(in_dst)
+    async fn handle_timeout(&self, token: u64) -> Result<Vec<Command>> {
+        match &mut *self.state.lock().await {
+            State::Approved(state) => state.handle_timeout(token),
+            _ => Ok(vec![]),
```

Comment:
other stages won't timeout?

Reply:
That's right. The bootstraping/joining timeout is better handled by the upper layers.